README.md
SpeakIt v0.4
=======================

This program reads the typed in text and searches a folder called 'words' for a sound file that corresponds to each individual word in the sentence. So the more sound files in the 'word' folder the more words that can be spoken from the input sentence. A sample collection of words can be downloaded from The Hailstone Web site (396kb)

## What's new in 0.4

- Source code now included.
- Cleaned up source a bit so should be easier to follow the code.


## Usage

 1.
  SpeakIt [-h] ["text"]

  Options:
    -h Does not show SpeakIt window
    "text" The text you want SpeakIt to say upon launch

  Example (in Terminal): SpeakIt -h "Welcome Be dude" &

 2.
  Click on the SpeakIt program

## Notes on source code

- The source is being released because Sean, the guy at Hailstone who wrote this app, does not have time to update it and does not want to see the app just die off.
- This program was originally written to learn the Be API and as such is not structured all that well.
- The PowerPC version would not compile with the cross platform tools, however the source is now available for someone to compile it on a PowerPC machine.

POSSIBLE ITEMS TO ADD:
   - Open plain text files and then speak.
   - A monitor mode that will import any new text put on the clipboard.
   - Drag and Drop support for text.
   - Allow user to change delay between words spoken.
   - Squash any bugs in the program.


## Source Code License

The included text and source code contained with this file may not be used for commercial purposes without prior authorization from Hailstone Software.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to use, copy, modify, merge, distribute, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software. Moreover, the name of any contributor to the Software shall remain unaltered in the source files.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL HAILSTONE SOFTWARE BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.




Hailstone Software
www.hailstonesoftware.com
send questions or comments to help@hailstonesoftware.com




CHANGELOG
SpeakIt 0.4 <February 14 2000>

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Source code now included.
- Cleaned up a bit so should be easier to follow the code.

//---------------------------------------------

SpeakIt 0.3 <July 24 1999>

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Redesigned the internal guts of the program, should be much more efficient now.
- Switched to the new BGameSound class which seemed to fix the strange sound playback
  in R4.5.
- Because we made the above switch R4.5 will be needed to run this app, unless we receive
  requests for a R4 version.
- Removed the 100 character limit, now can change it from the settings menu.

//---------------------------------------------

SpeakIt 0.2 <May 30 1999>
(First beta release) 

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Fixed problem with line feed character preventing last word in sentence from
  being played when user presses enter at end.
- Was including a lib that was not needed so the ram usage of the app went 
  down by a megabyte or two.
- Finally got around to including slightly better documentation. :) 

//---------------------------------------------

SpeakIt 0.1.7 Alpha <May 09 1999>
(Last alpha release) 

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Window position now saved when program shuts down.
- Anything that is not a number or letter is stripped away from text when 
  spoken.

//---------------------------------------------

SpeakIt 0.1.6 Alpha <April 02 1999>

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Fixed a bug in the sound code that caused a memory leak. Thanks to Sean 
  Heber for pointing this out.
- Added the -h (or -H) command line option to hide the window when launching, 
  very useful for playing startup sounds.
  EXAMPLE:	SpeakIt -h "Welcome be dude" &

* KNOWN PROBLEMS / LIMITATIONS *
- Does not remember previous window positions.

//---------------------------------------------

SpeakIt 0.1.5 Alpha <March 25 1999>

* FIXED IN THIS RELEASE / OR ADDED FEATURES *
- Now remembers previous font size or color.
- Can now play any sound file format supported by the BeOS and does not have
  to use only 44.1 khz 8 / 16 bit stereo wave files. The draw back to this change
  is you have to change any "sound.wav" files to "sound" (must be lower case).
- You now have the ability to chose any font color with a color control panel.
- Now supports multiple launches of the program.
- You can now launch the app from a command line and pass it a string of text.
  EXAMPLE:	SpeakIt "Welcome be dude" &
- The cripple limit of letters is now set to 100 instead of 50.

* KNOWN PROBLEMS / LIMITATIONS *
- Does not remember previous window positions.

//---------------------------------------------

SpeakIt 0.1.4 Alpha <March 15 1999>

* KNOWN PROBLEMS *
- Does not remember previous font size or color.
- Does not remember previous window positions.
- Still is limited to 44.1 khz 8 / 16 bit stereo wav files.
- Any wave files you add must be lowercase and have the .wav extension.


source/AppWindow.cpp
// Original Author: Sean Long @ Hailstone Software
// Contact: 		sean@hailstonesoftware.com
// Last Modified:   02/14/2000
#include <Application.h>
#include <MenuBar.h>
#include <MenuItem.h>
#include <Button.h>
#include <ScrollView.h>
#include <TextView.h>

//For string manipulation
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "MsgVals.h"
#include "Colors.h"
#include "AppWindow.h"
#include "SoundWork.h"

#include "SettingsClass.h"

//Global settings
extern SettingsClass		settings;


AppWindow::AppWindow(char *StartText,bool Hide,BRect Frame)
			: BWindow(Frame, "SpeakIt", B_DOCUMENT_WINDOW, B_NOT_ZOOMABLE,B_CURRENT_WORKSPACE)
{			

	BMenu 			*tempMenu;
	BMenu 			*sizemenu;
	BMenuItem 		*tempMenuItem;
	BMenuItem 		*font,*color;
	
	MoveTo(settings.TopLeft);
	
	appView = new BView(Bounds(), "SpeakItView", B_FOLLOW_ALL, B_WILL_DRAW);
	appView->SetViewColor(BeBackgroundGrey);
	AddChild(appView);

	
	strcpy(totalpath,settings.SoundPath);
	//not currently used for anything
	strcpy(CDpath,"/SPEAKIT/WORDS/");
	

	// creating and adding menu bar
	MenuRect = Bounds();
	MenuRect.bottom = 19;
	MenuBar = new BMenuBar(MenuRect, "Menu", B_FOLLOW_LEFT_RIGHT|B_FOLLOW_TOP, B_ITEMS_IN_ROW, true);


	//Layout the File menu
	tempMenu = new BMenu("File");
	
	tempMenuItem = new BMenuItem("About SpeakIt", new BMessage(B_ABOUT_REQUESTED));
    tempMenuItem->SetTarget(be_app); 
	tempMenu->AddItem(tempMenuItem);
	tempMenu->AddSeparatorItem();
	tempMenuItem = new BMenuItem("Quit", new BMessage(B_QUIT_REQUESTED),'Q');
	tempMenuItem->SetTarget(be_app); 
	tempMenu->AddItem(tempMenuItem);
	MenuBar->AddItem(tempMenu);

	//Layout the Edit menu
	tempMenu = new BMenu("Edit");

	tempMenuItem = new BMenuItem("Cut", new BMessage(B_CUT),'X');
	tempMenu->AddItem(tempMenuItem);
	tempMenuItem = new BMenuItem("Copy", new BMessage(B_COPY),'C');
	tempMenu->AddItem(tempMenuItem);
	tempMenuItem = new BMenuItem("Paste", new BMessage(B_PASTE),'V');
	tempMenu->AddItem(tempMenuItem);
	tempMenuItem = new BMenuItem("Clear", new BMessage(SPEAKIT_CLEAR));
	tempMenu->AddItem(tempMenuItem);
	tempMenu->AddSeparatorItem();
	tempMenuItem = new BMenuItem("Select All", new BMessage(B_SELECT_ALL),'A');
	tempMenu->AddItem(tempMenuItem);
	MenuBar->AddItem(tempMenu);
	
	//Layout the Font menu
	tempMenu = new BMenu("Font");	
	tempMenu->AddItem(new BMenuItem(sizemenu = new BMenu("Size"), NULL));
	
	sizemenu->AddItem(font = new BMenuItem("10", new BMessage(SPEAKIT_FONT_10)));
	if(settings.FontSize == 10)
		font->SetMarked(true);
	sizemenu->AddItem(font = new BMenuItem("12", new BMessage(SPEAKIT_FONT_12)));
	if(settings.FontSize == 12)
		font->SetMarked(true);
	sizemenu->AddItem(font = new BMenuItem("14", new BMessage(SPEAKIT_FONT_14)));
	if(settings.FontSize == 14)
		font->SetMarked(true);
	sizemenu->AddItem(font = new BMenuItem("18", new BMessage(SPEAKIT_FONT_18)));
	if(settings.FontSize == 18)
		font->SetMarked(true);
	sizemenu->AddItem(font = new BMenuItem("24", new BMessage(SPEAKIT_FONT_24)));
	if(settings.FontSize == 24)
		font->SetMarked(true);
	sizemenu->AddItem(font = new BMenuItem("36", new BMessage(SPEAKIT_FONT_36)));
	if(settings.FontSize == 36)
		font->SetMarked(true);
	sizemenu->SetRadioMode(true);

	tempMenu->AddSeparatorItem();
	color = new BMenuItem("Color", new BMessage(SPEAKIT_COLOR_OTHER));
	color->SetTarget(be_app);
	tempMenu->AddItem(color);
	MenuBar->AddItem(tempMenu);
	
	//Layout the Window menu
	tempMenu = new BMenu("Window");	
	tempMenuItem = new BMenuItem("Settings", new BMessage(SPEAKIT_SETTINGS));
	tempMenuItem->SetTarget(be_app); 
	tempMenu->AddItem(tempMenuItem);
	MenuBar->AddItem(tempMenu);
	
	//Now create the menu
	appView->AddChild(MenuBar);
	
	// Add the CUT button	//1 is a spacer  //40 is width of the button  
	ButtonRect.Set(0,MenuRect.bottom+1,40,MenuRect.bottom+25);	
	appView->AddChild(button = new BButton(ButtonRect, "Cut", "Cut", new BMessage(B_CUT), B_FOLLOW_NONE));	
	// Add the COPY button	  
	ButtonRect.Set(ButtonRect.right +1,MenuRect.bottom+1,ButtonRect.right +40,MenuRect.bottom+25);
	appView->AddChild(button = new BButton(ButtonRect, "Copy", "Copy", new BMessage(B_COPY), B_FOLLOW_NONE));	
	// Add the PASTE button	
	ButtonRect.Set(ButtonRect.right +1,MenuRect.bottom+1,ButtonRect.right +40,MenuRect.bottom+25);
	appView->AddChild(button = new BButton(ButtonRect, "Paste", "Paste", new BMessage(B_PASTE), B_FOLLOW_NONE));	
	// Add the CLEAR button	 
	ButtonRect.Set(ButtonRect.right +1,MenuRect.bottom+1,ButtonRect.right +40,MenuRect.bottom+25);
	appView->AddChild(button = new BButton(ButtonRect, "Clear", "Clear", new BMessage(SPEAKIT_CLEAR), B_FOLLOW_NONE));		
	// Add the SELECT ALL button  						60 is width of the button  
	ButtonRect.Set(ButtonRect.right +1,MenuRect.bottom+1,ButtonRect.right +60,MenuRect.bottom+25);
	appView->AddChild(button = new BButton(ButtonRect, "Select All", "Select All", new BMessage(B_SELECT_ALL), B_FOLLOW_NONE));			
	// Add the SPEAK button	  
	ButtonRect.Set(Frame.right-50,MenuRect.bottom+1,Frame.right-6,MenuRect.bottom+25);
	appView->AddChild(button = new BButton(ButtonRect, "Speak", "Speak", new BMessage(SPEAKIT_WORK), B_FOLLOW_RIGHT));	
	
	// Add the text view
	textfont.SetSize(settings.FontSize);				// 25 is height of menu
	TextFrame.Set(0,ButtonRect.bottom+3,Frame.right-20,Frame.bottom-25);
	TextRect.Set(0,0,Frame.right-B_V_SCROLL_BAR_WIDTH-3,Frame.bottom);		
	textview = new BTextView(TextFrame, "text_view", TextRect, B_FOLLOW_ALL_SIDES, B_WILL_DRAW);	
	textview->SetDoesUndo(true);

	textview->SetFontAndColor(&textfont,B_FONT_ALL,&settings.FontColor);

	textview->SetMaxBytes(settings.MaxTextLength);
	
				
	// Add the scroll view
	appView->AddChild(scrollview = new BScrollView("scroll_view", textview, B_FOLLOW_ALL_SIDES, B_FRAME_EVENTS, false, true, B_FANCY_BORDER));
	scrollview->AttachedToWindow();

	textview->MakeFocus(true);
	
	textview->SetText(StartText);
		
	//Say start up text if it is pressent
	Work();
	if(Hide==true)
		be_app->PostMessage(B_QUIT_REQUESTED);

}// end SpeakItWindow

//////////////////////////////////////////////////////////////////////////

void AppWindow::MessageReceived(BMessage *message)
{
	
	int32			insertstart,insertfinish;

	
	switch ( message->what )
	{

//App specific messages
		case SPEAKIT_WORK:
			 Work();
			 break;
		case SPEAKIT_CLEAR:
			 textview->SetText("");
			 break;	
		case B_SELECT_ALL:
			 textview->SelectAll();
			 break;				
		case SPEAKIT_FONT_10:		
			 textfont.SetSize(10.0);
			 textview->SetFontAndColor(&textfont);
			 break;	
		case SPEAKIT_FONT_12:
			 textfont.SetSize(12.0);
			 textview->SetFontAndColor(&textfont);
			 break;
		case SPEAKIT_FONT_14:
			 textfont.SetSize(14.0);
			 textview->SetFontAndColor(&textfont);
			 break;
		case SPEAKIT_FONT_18:
			 textfont.SetSize(18.0);
			 textview->SetFontAndColor(&textfont);
			 break;
		case SPEAKIT_FONT_24:
			 textfont.SetSize(24.0);
			 textview->SetFontAndColor(&textfont);
			 break;
		case SPEAKIT_FONT_36:
			 textfont.SetSize(36.0);
			 textview->SetFontAndColor(&textfont);
		     break;
		 
		case SPEAKIT_COLORWIN_CHANGE:		 
			 //Gets currnet selection and the turns text into new color and reselects anything selected
			 textview->GetSelection(&insertstart,&insertfinish);
			 textview->SetFontAndColor(&textfont,B_FONT_ALL,&settings.FontColor);
			 textview->SetText(textview->Text());
			 textview->Select(insertstart,insertfinish);		
			 break;
			 
		case SPEAKIT_CHAR_LEN_CHANGE:
			 textview->SetMaxBytes(settings.MaxTextLength);
			 break;
			 
//General messages	
		case B_CUT:
			 textview->Cut(be_clipboard);
			 break;
		case B_COPY:
			 textview->Copy(be_clipboard);
			 break;
		case B_PASTE:
			 textview->Paste(be_clipboard);
			 break;
		case B_QUIT_REQUESTED:			 
			 be_app->PostMessage(B_QUIT_REQUESTED);
			 break;
		
   		default:
			 BWindow::MessageReceived(message);
			 break;
	}
		
	
}// end MessageReceived

//////////////////////////////////////////////////////////////////////////


bool AppWindow::QuitRequested()
{	
	
	BRect			top_left;
	
	//Setting the font size	
	settings.FontSize = (int8) textfont.Size();
	
	//Setting the top left hand point of the current window position
	top_left = Frame();
	settings.TopLeft.x = top_left.left;
	settings.TopLeft.y = top_left.top;

	be_app -> PostMessage(B_QUIT_REQUESTED);
		
	return(true);

}


//////////////////////////////////////////////////////////////////////////


void AppWindow::Work()
{
// Function where all the text to wav stuff is done
	char *word;
	
	char *sentance;
	
	char tmppath[B_FILE_NAME_LENGTH+B_FILE_NAME_LENGTH];
	int isnumber=false;
	int index=0;
	int32 currentoff=0;
	int32 startoff=0;
	int32 endoff=0;
	int32 previousstart=0,previousfinish=0;	
	
	textview->SetMaxBytes(settings.MaxTextLength);
	
	sentance = new char[textview->TextLength()]; 
	
	strcpy(sentance,textview->Text());
	
	word = strtok(sentance," ");
	//say word
	if(word) //only run if text entered
	do
	{
		isnumber = false;
		//say word
	
		for(index=0;index <(signed)strlen(word);index++)
		{
			word[index]=tolower(word[index]);		//make lower case
		
			//Turn on the isnumber flag if word is a number (or contains a number)
			switch (word[index])
			{
				case '0':	
				case '1': 
				case '2':
				case '3':
				case '4':
				case '5':
				case '6':
				case '7':
				case '8':
				case '9': isnumber=true;
						  break;
				default : break;
			}//end switch			
	
		}

		if(settings.Highlight)
		{
			//get start and end offest of current word
			//highlight that word
			textview->FindWord(currentoff,&startoff,&endoff);
			
			textview->Highlight(previousstart,previousfinish);
			textview->Draw(TextRect);
			textview->Highlight(startoff,endoff);
			textview->Draw(TextRect);
			currentoff = endoff + 1;
			previousstart = startoff; 
			previousfinish= endoff;
		
		}
		
		if(isnumber ==false)
		{	
			RemovePuncFromWord(word,strlen(word));
			//Try playing the sound from the local words dir
			strcpy(tmppath,totalpath);
			strcat(tmppath,word);			
			if(playsound(tmppath,settings.Volume) == false )
			{
				//Play from the CD if word not in local dir
				strcpy(tmppath,CDpath);
							
				//make path UPPER case for CD
				for(index=0;index< (signed)strlen(word);index++)
					word[index]=toupper(word[index]);
							
				strcat(tmppath,word);		
								
				playsound(tmppath,settings.Volume);
						
			}
		
		}
		else
		{
			//word IS numeric so process it	
			RemovePuncFromWord(word,strlen(word));	//removes $ , .

			strcpy(tmppath,totalpath);
			ProcessNum(tmppath,atoi(word),strlen(word),settings.Volume);
					
		}
					
		
	}while((word = strtok(NULL," ")));
	
	//clean up any leftover highlihting
	textview->Invalidate();
	
	delete sentance;

}//end Work()

//////////////////////////////////////////////////////////////////////////

void AppWindow::FrameResized(float w, float h)
{
	if(w!=-1 && h!=-1) BWindow::FrameResized(w,h);

		BRect rect;
		if(w==-1 && h==-1)
		{
			rect = Bounds();
			h = rect.bottom;
			w = rect.right;
		}
		else
		{
		rect = textview->Frame();
		rect.bottom = rect.top + h -B_H_SCROLL_BAR_HEIGHT;
		rect.right = rect.left + w -B_V_SCROLL_BAR_WIDTH-1;
		textview->SetTextRect(rect);
		}
	
}

//////////////////////////////////////////////////////////////////////////
